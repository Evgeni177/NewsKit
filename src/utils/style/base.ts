import {Theme, BreakpointKeys} from '../../theme';
import {isResponsive, getMediaQueryFromTheme} from '../responsive-helpers';
import {filterObject} from '../filter-object';
import {getToken} from '../get-token';
import {ThemeProp} from '../style-types';
import {MQ} from './types';

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const isValid = (value: any): boolean =>
  !(Number.isNaN(value) || Array.isArray(value) || typeof value === 'object');

export const getDefaultedValue = <
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  FromThemeUtil extends (...args: any) => any
>(
  getPresetFromThemeUtil: FromThemeUtil,
  presetType: string,
  cssProp?: string,
) => <Props extends ThemeProp>(
  defaultPath: string | undefined,
  overridePath: string | false = '',
  option?: Parameters<FromThemeUtil>[2],
) => (props: Props) =>
  getPresetFromThemeUtil(
    getToken(props, defaultPath, overridePath, presetType),
    undefined,
    cssProp || option,
  )(props);

export const getValueFromTheme = <ThemeToken extends string>(
  themeKey: keyof Theme,
) => <Props extends ThemeProp>(
  defaultToken?: MQ<ThemeToken>,
  customProp?: Exclude<keyof Props, 'theme'>,
  allowNonThemeValue?: boolean,
) => (props: Props) => {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const section = props.theme[themeKey] as any;
  const token = (customProp && props[customProp]) || defaultToken;

  return (
    (section && section[token]) ||
    (allowNonThemeValue && isValid(token) ? token : '')
  );
};

export const getResponsiveValueFromTheme = <ThemeToken extends string>(
  themeKey: keyof Theme,
) => <Props extends ThemeProp>(
  defaultToken?: MQ<ThemeToken>,
  customProp?: Exclude<keyof Props, 'theme'>,
) => ({theme, ...props}: Props) => {
  const section = theme[themeKey] as Record<ThemeToken, unknown>;
  const propKeys = (customProp && props[customProp]) || defaultToken;
  const {breakpoints} = theme;
  if (isResponsive(propKeys, breakpoints)) {
    // We have a breakpoints object...

    // Convert breakpoints to array and order them
    const breakpointKeys = Object.keys(breakpoints) as BreakpointKeys[];
    const mq = breakpointKeys.sort((a, b) => breakpoints[a] - breakpoints[b]);

    const presetKeys: [BreakpointKeys, ThemeToken][] = Object.entries(
      filterObject(propKeys, mq),
    ).sort(
      ([a], [b]) =>
        mq.indexOf(a as BreakpointKeys) - mq.indexOf(b as BreakpointKeys),
    ) as any; // eslint-disable-line @typescript-eslint/no-explicit-any

    const cssObject = presetKeys
      .filter(
        // Exclude invalid breakpoints and theme section keys
        ([breakpointKey, presetKey]) =>
          presetKey &&
          section[presetKey] &&
          breakpointKeys.includes(breakpointKey),
      )
      .reduce(
        (acc, [key, presetKey], index, arr) => {
          const preset = section[presetKey];
          // Get next key to set the max. This stops styles overlapping when they
          // shouldn't by explicitly setting them for the range they need to apply on.
          const nextKey = arr[index + 1] ? arr[index + 1][0] : undefined;
          const mediaQuery = getMediaQueryFromTheme(key, nextKey)({theme});
          acc[mediaQuery] = preset;
          return acc;
        },
        {} as Record<string, unknown>,
      );
    return Object.entries(cssObject);
  }

  return (propKeys && section[propKeys as ThemeToken]) || '';
};
