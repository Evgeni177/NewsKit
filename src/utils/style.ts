import baseStyled, {CreateStyled, CSSObject} from '@emotion/styled';
import {css} from '@emotion/core';
import {
  TypePresets,
  TypePresetKeys,
  ColorKeys,
  AnimationKeys,
  BorderKeys,
  Theme,
  FontPrimitivesKeys,
  BreakpointKeys,
} from '../themes';
import {FontSizeKeys, LineHeightKeys} from '../themes/newskit-light/fonts';
import {getFontSizing} from './font-sizing';
import {GridKeys} from '../themes/newskit-light/grid';
import {ShadowKeys} from '../themes/newskit-light/shadow';
import {SizingKeys, IconSizeKeys} from '../themes/newskit-light/spacing';
import {MarginPresetKeys, PaddingPresetKeys} from '../themes/mappers/spacing';
import {BorderRadiusKeys} from '../themes/mappers/border-radius';
import {getMediaQuery, isResponsive} from './responsive-helpers';
import {filterObject} from './filter-object';
import {isFontConfigObject} from './guards';
import {getFontProps} from './get-font-props';
import {getToken} from './get-token';
import {ThemeProp} from './style-types';

export {CSSObject} from '@emotion/core';

export {css};

// Cast styled with the Theme so we don't have to specify theme at every usage.
export const styled = baseStyled as CreateStyled<Theme>;

export type Responsive<T> = T | T[];
export type MQ<T> =
  | T
  | Partial<{
      xs: T;
      sm: T;
      md: T;
      lg: T;
    }>;

export const getDefaultedValue = <
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  FromThemeUtil extends (...args: any) => any
>(
  getPresetFromThemeUtil: FromThemeUtil,
  presetType: string,
) => <Props extends ThemeProp>(
  defaultPath: string,
  overridePath: string | false = '',
  option?: Parameters<FromThemeUtil>[2],
) => (props: Props) =>
  getPresetFromThemeUtil(
    getToken(props, defaultPath, overridePath, presetType),
    undefined,
    option,
  )(props);

export const getValueFromTheme = <ThemeToken extends string>(
  themeKey: keyof Theme,
) => <Props extends ThemeProp>(
  defaultToken?: MQ<ThemeToken>,
  customProp?: Exclude<keyof Props, 'theme'>,
) => ({theme, ...props}: Props) => {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const section = theme[themeKey] as any;
  const propKeys = (customProp && props[customProp]) || defaultToken;
  return (propKeys && section[propKeys]) || '';
};

export const getResponsiveValueFromTheme = <ThemeToken extends string>(
  themeKey: keyof Theme,
) => <Props extends ThemeProp>(
  defaultToken?: MQ<ThemeToken>,
  customProp?: Exclude<keyof Props, 'theme'>,
) => ({theme, ...props}: Props) => {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const section = theme[themeKey] as any;
  const propKeys = (customProp && props[customProp]) || defaultToken;
  const {breakpoints} = theme;
  if (
    propKeys &&
    typeof propKeys === 'object' &&
    isResponsive(propKeys, breakpoints)
  ) {
    const mq = (Object.keys(breakpoints) as BreakpointKeys[]).sort(
      (a, b) => breakpoints[a] - breakpoints[b],
    );
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const presetKeys = Object.entries(filterObject(propKeys, mq as any)).sort(
      ([a], [b]) =>
        mq.indexOf(a as BreakpointKeys) - mq.indexOf(b as BreakpointKeys),
    );
    const cssObject = presetKeys.reduce(
      (acc, [key, presetKey]) => {
        const preset = section[presetKey as ThemeToken];
        if (preset) {
          const mediaQuery = getMediaQuery({
            'min-width': `${breakpoints[key as BreakpointKeys]}px`,
          });
          acc[mediaQuery] = preset;
        }
        return acc;
      },
      {} as CSSObject,
    );
    return Object.entries(cssObject);
  }

  return (propKeys && section[propKeys]) || '';
};

// TODO: (https://nidigitalsolutions.jira.com/browse/PPDSC-1143) renaming this could be a breaking change to any consumers outside NewsKit using this. Remove this and fix imports when ready to make the change.
export const getPresetValueFromTheme = getResponsiveValueFromTheme;

export const getTypePresetFromTheme = <Props extends ThemeProp>(
  defaultToken?: MQ<TypePresetKeys>,
  customProp?: Exclude<keyof Props, 'theme'>,
  options?: {withCrop: boolean},
) => (props: Props) => {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const applyCrop = (typePreset: any) => {
    const {fontSize, lineHeight, fontFamily} = typePreset;
    const cropProps = getFontProps(fontSize, lineHeight, fontFamily);
    return {
      ...typePreset,
      ...cropProps,
    };
  };

  const {withCrop = false} = options || {};
  const typePreset = getResponsiveValueFromTheme('typePresets')(
    defaultToken,
    customProp,
  )(props) as Partial<TypePresets[TypePresetKeys]> | Array<[string, CSSObject]>;

  if (Array.isArray(typePreset)) {
    return typePreset.reduce(
      (acc, [mq, cssObject], index) => {
        let cssObjectFinal = cssObject;

        if (withCrop && !Array.isArray(cssObject)) {
          cssObjectFinal = applyCrop(cssObject);
        }

        if (index === 0) {
          return {...acc, ...cssObjectFinal};
        }
        acc[mq] = cssObjectFinal;
        return acc;
      },
      {} as CSSObject,
    );
  }

  if (withCrop && !Array.isArray(typePreset)) {
    return applyCrop(typePreset);
  }
  return typePreset;
};

export const getTypePreset = getDefaultedValue(
  getTypePresetFromTheme,
  'typePreset',
);

export const getAnimationFromTheme = getValueFromTheme<AnimationKeys>(
  'animation',
);

export const getFontsFromTheme = <Props extends ThemeProp>(
  defaultToken?: MQ<FontPrimitivesKeys>,
  customProp?: Exclude<keyof Props, 'theme'>,
) => (props: Props) => {
  const section = props.theme.fonts;
  const propKeys = (customProp && props[customProp]) || defaultToken;
  const style = section[propKeys as FontPrimitivesKeys];

  if (style && isFontConfigObject(style)) {
    return style.fontFamily;
  }

  return propKeys ? style : '';
};

export const getColorFromTheme = getValueFromTheme<ColorKeys>('colors');

export const getSizingFromTheme = getValueFromTheme<
  SizingKeys | MarginPresetKeys | PaddingPresetKeys | IconSizeKeys
>('sizing');

export const getSizing = getDefaultedValue(getSizingFromTheme, 'sizing');

export const getPaddingPresetFromTheme = <Props extends ThemeProp>(
  defaultToken?: MQ<PaddingPresetKeys>,
  customProp?: Exclude<keyof Props, 'theme'>,
) => (props: Props) => {
  const padding = getResponsiveValueFromTheme('sizing')(
    defaultToken,
    customProp,
  )(props) as string | Array<[string, string]>;
  if (Array.isArray(padding)) {
    return padding.reduce(
      (acc, [mq, preset], index) => {
        const style = {padding: preset};
        if (index === 0) {
          return style;
        }
        acc[mq] = style;
        return acc;
      },
      {} as CSSObject,
    );
  }

  return {padding};
};

export const getPaddingPreset = getDefaultedValue(
  getPaddingPresetFromTheme,
  'paddingPreset',
);

export const getMarginPresetFromTheme = <Props extends ThemeProp>(
  defaultToken?: MQ<MarginPresetKeys>,
  customProp?: Exclude<keyof Props, 'theme'>,
) => (props: Props) => {
  const margin = getResponsiveValueFromTheme('sizing')(
    defaultToken,
    customProp,
  )(props) as string | Array<[string, string]>;
  if (Array.isArray(margin)) {
    return margin.reduce(
      (acc, [mq, preset], index) => {
        const style = {margin: preset};
        if (index === 0) {
          return style;
        }
        acc[mq] = style;
        return acc;
      },
      {} as CSSObject,
    );
  }

  return {margin};
};

export const getMarginPreset = getDefaultedValue(
  getMarginPresetFromTheme,
  'marginPreset',
);

export const getBorderFromTheme = getValueFromTheme<BorderKeys>('borders');

export const getBorderRadiusFromTheme = getValueFromTheme<BorderRadiusKeys>(
  'borderRadius',
);

export const getGridSettingFromTheme = getValueFromTheme<GridKeys>('grid');

export const getShadowFromTheme = getValueFromTheme<ShadowKeys>('shadow');

export const getFontSizingFromTheme = (
  fontSizeKey: TypePresetKeys | FontSizeKeys,
  lineHeightKey: LineHeightKeys,
) => ({theme}: ThemeProp) => {
  const typePresets = theme.typePresets[fontSizeKey as TypePresetKeys];
  const fontSize = typePresets
    ? typePresets.fontSize
    : theme.fonts[fontSizeKey as FontSizeKeys];
  return getFontSizing(fontSize, theme.fonts[lineHeightKey]);
};
