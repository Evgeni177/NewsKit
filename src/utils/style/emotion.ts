/* istanbul ignore file */
import '@emotion/react';
import styled from '@emotion/styled';

import {Theme as DefaultTheme} from '../../theme';

declare module '@emotion/react' {
  export interface Theme extends DefaultTheme {}
}

export type {CSSObject, SerializedStyles} from '@emotion/react';
export {Global, css, keyframes} from '@emotion/react';
export {styled};
