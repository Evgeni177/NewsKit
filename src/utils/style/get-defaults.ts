import {getDefaultedValue} from './base';
import {
  getBorderFromTheme,
  getSizingFromTheme,
  getSpacingFromTheme,
  getSpacingInsetFromTheme,
  getTypographyPresetFromTheme,
} from './getters';

export const getTypographyPreset = getDefaultedValue(
  getTypographyPresetFromTheme,
  'typographyPreset',
);

export const getSpacingInlineHorizontal = getDefaultedValue(
  getSpacingFromTheme,
  'spaceInline',
  'marginRight',
);

export const getSpacingInlineVertical = getDefaultedValue(
  getSpacingFromTheme,
  'spaceInline',
  'marginBottom',
);
export const getSpacingStackHorizontal = getDefaultedValue(
  getSpacingFromTheme,
  'spaceStack',
  'marginBottom',
);

export const getSpacingStackVertical = getDefaultedValue(
  getSpacingFromTheme,
  'spaceStack',
  'marginRight',
);

export const getSpacingInset = getDefaultedValue(
  getSpacingInsetFromTheme,
  'spaceInset',
);

// TODO: Which of these funcs is no longer used ?
export const getSizing = getDefaultedValue(getSizingFromTheme, 'sizing');
export const getMinHeight = getDefaultedValue(getSizingFromTheme, 'minHeight');
export const getMinWidth = getDefaultedValue(getSizingFromTheme, 'minWidth');
export const getHeight = getDefaultedValue(getSizingFromTheme, 'height');
export const getWidth = getDefaultedValue(getSizingFromTheme, 'width');
export const getSize = getDefaultedValue(getSizingFromTheme, 'size');
export const getMaxSize = getDefaultedValue(getSizingFromTheme, 'maxSize');
export const getMinSize = getDefaultedValue(getSizingFromTheme, 'minSize');
export const getSpace = getDefaultedValue(getSpacingFromTheme, 'space');
export const getWeight = getDefaultedValue(getBorderFromTheme, 'weight');
