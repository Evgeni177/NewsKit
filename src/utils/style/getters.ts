import {CSSObject} from '@emotion/styled';
import {getFontSizing} from '../font-sizing';
import {TypographyPreset} from '../../theme';
import {isFontConfigObject} from '../guards';
import {getFontProps} from '../get-font-props';
import {ThemeProp} from '../style-types';
import {MQ} from './types';
import {getResponsiveValueFromTheme, getValueFromTheme} from './base';

export const getTypographyPresetFromTheme = <Props extends ThemeProp>(
  defaultToken?: MQ<string>,
  customProp?: Exclude<keyof Props, 'theme'>,
  options?: {withCrop: boolean},
) => (props: Props) => {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const applyCrop = (typographyPreset: any) => {
    const {fontSize, lineHeight, fontFamily} = typographyPreset;
    const cropProps = getFontProps(
      fontSize,
      lineHeight,
      fontFamily,
      props.theme.fonts,
    );
    return {
      ...typographyPreset,
      ...cropProps,
    };
  };

  const {withCrop = false} = options || {};
  const typographyPreset = getResponsiveValueFromTheme('typographyPresets')(
    defaultToken,
    customProp,
  )(props) as Partial<TypographyPreset> | Array<[string, CSSObject]>;

  if (Array.isArray(typographyPreset)) {
    return typographyPreset.reduce(
      (acc, [mq, cssObject]) => {
        let cssObjectFinal = cssObject;

        if (withCrop && !Array.isArray(cssObject)) {
          cssObjectFinal = applyCrop(cssObject);
        }

        acc[mq] = cssObjectFinal;
        return acc;
      },
      {} as CSSObject,
    );
  }

  if (typographyPreset && withCrop && !Array.isArray(typographyPreset)) {
    return applyCrop(typographyPreset);
  }
  return typographyPreset;
};

export const getMotionFromTheme = getValueFromTheme<string>('motions');

export const getFontsFromTheme = <Props extends ThemeProp>(
  defaultToken?: MQ<string>,
  customProp?: Exclude<keyof Props, 'theme'>,
) => (props: Props) => {
  const section = props.theme.fonts;
  const propKeys = (customProp && props[customProp]) || defaultToken;
  const style = section[propKeys as string];

  if (style && isFontConfigObject(style)) {
    return style.fontFamily;
  }

  return propKeys ? style : '';
};

export const getColorFromTheme = getValueFromTheme<string>('colors');

export const getSizingFromTheme = getValueFromTheme<string>('sizing');

export const getSpacingFromTheme = <Props extends ThemeProp>(
  defaultToken?: MQ<string>,
  customProp?: Exclude<keyof Props, 'theme'>,
  cssProp?: string,
) => (props: Props) => {
  const value = getResponsiveValueFromTheme('spacePresets')(
    defaultToken,
    customProp,
  )(props) as string | Array<[string, string]>;
  if (Array.isArray(value)) {
    return value.reduce(
      (acc, [mq, preset]) => {
        acc[mq] = cssProp && {[cssProp]: preset};
        return acc;
      },
      {} as CSSObject,
    );
  }

  return cssProp ? {[cssProp]: value} : value;
};

export const getSpacingInsetFromTheme = <Props extends ThemeProp>(
  defaultToken?: MQ<string>,
  customProp?: Exclude<keyof Props, 'theme'>,
) => (props: Props) => {
  const padding = getResponsiveValueFromTheme('spacePresets')(
    defaultToken,
    customProp,
  )(props) as string | Array<[string, string]>;
  if (Array.isArray(padding)) {
    return padding.reduce(
      (acc, [mq, preset]) => {
        acc[mq] = {padding: preset};
        return acc;
      },
      {} as CSSObject,
    );
  }

  return {padding};
};

// Legacy function, clients might be using it, do not use internally
export const getMarginPresetFromTheme = <Props extends ThemeProp>(
  defaultToken?: MQ<string>,
  customProp?: Exclude<keyof Props, 'theme'>,
) => (props: Props) => {
  const margin = getResponsiveValueFromTheme('spacePresets')(
    defaultToken,
    customProp,
  )(props) as string | Array<[string, string]>;
  if (Array.isArray(margin)) {
    return margin.reduce(
      (acc, [mq, preset]) => {
        acc[mq] = {margin: preset};
        return acc;
      },
      {} as CSSObject,
    );
  }

  return {margin};
};

export const getBorderFromTheme = getValueFromTheme<string>('borders');

export const getBorderRadiusFromTheme = getValueFromTheme<string>('borders');

export const getShadowFromTheme = getValueFromTheme<string>('shadows');

export const getFontSizingFromTheme = (
  fontSizeKey: string,
  lineHeightKey: string,
) => ({theme}: ThemeProp) => {
  const typographyPresets = theme.typographyPresets[fontSizeKey];
  const fontSize = typographyPresets
    ? typographyPresets.fontSize
    : theme.fonts[fontSizeKey];
  return getFontSizing(fontSize, theme.fonts[lineHeightKey]);
};
