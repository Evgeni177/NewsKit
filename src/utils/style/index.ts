export * from './base';
export * from './emotion';
export * from './getters';
export * from './style-preset';
export * from './types';
export * from '../deep-merge';
