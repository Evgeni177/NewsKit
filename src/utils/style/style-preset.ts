import {
  Theme,
  StylePresetStates,
  StylePresetStyleKeys,
  StylePresetStyles,
  StylePreset,
} from '../../theme';
import {filterObject, rejectObject} from '../filter-object';
import {ThemeProp} from '../style-types';
import {getDefaultedValue, getResponsiveValueFromTheme} from './base';
import {CSSObject} from './emotion';
import {MQ} from './types';

export interface GetStylePresetFromThemeOptions {
  nestedCssSelector?: string;
  isLoading?: boolean;
  isSelected?: boolean;
  isDisabled?: boolean;
  isInvalid?: boolean;
  isValid?: boolean;
  isSvg?: boolean;
  omitStates?: StylePresetStates[];
  filterStates?: StylePresetStates[];
  omitStyles?: StylePresetStyleKeys[];
  filterStyles?: StylePresetStyleKeys[];
}

/* When we are not on directly on a svg we need to add an 
  additional css selector to increase the specificity, allowing
  us to overrule the default color. Icon button is an example 
  of this in action */
const getCssSvgFillObject = (iconColor: string, isSvg: boolean): object =>
  isSvg
    ? {fill: iconColor}
    : {
        svg: {
          fill: iconColor,
        },
      };

export const getPresetStyles = (
  presetStyles: StylePresetStyles,
  options?: GetStylePresetFromThemeOptions,
) => {
  const {filterStyles = null, omitStyles = [], isSvg: isSvg = false} =
    options || {};
  const {iconColor, placeholderColor, ...cssObject} = filterStyles
    ? filterObject(presetStyles, filterStyles)
    : rejectObject(presetStyles, omitStyles);
  if (iconColor) {
    return {
      ...cssObject,
      ...getCssSvgFillObject(iconColor, isSvg),
    } as CSSObject;
  }

  if (placeholderColor) {
    return {
      ...cssObject,
      '::placeholder': {
        color: placeholderColor,
      },
    } as CSSObject;
  }

  return cssObject as CSSObject;
};

const getPresetStates = (
  stylePreset: StylePreset,
  options?: GetStylePresetFromThemeOptions,
) => {
  const {
    omitStates = [],
    filterStates = [],
    isSelected = false,
    isLoading = false,
    isDisabled = false,
    isInvalid = false,
    isValid = false,
  } = options || {};
  const {selected, loading, invalid, valid, ...presetStates} =
    filterStates && filterStates.length
      ? filterObject(stylePreset, filterStates)
      : rejectObject(stylePreset, omitStates);
  const stateOverrides =
    (isDisabled && presetStates.disabled) ||
    (isLoading && loading) ||
    (isSelected && selected) ||
    (isInvalid && invalid) ||
    (isValid && valid) ||
    undefined;
  if (stateOverrides) {
    const {base = {}} = presetStates;
    return {base: {...base, ...stateOverrides}};
  }

  return presetStates;
};

const getStylePresetValueFromTheme = (
  stylePreset: StylePreset,
  options?: GetStylePresetFromThemeOptions,
) =>
  Object.entries(getPresetStates(stylePreset, options)).reduce(
    (acc, [stateKey, presetState]) => {
      if (presetState) {
        const nestedCssSelector = options && options.nestedCssSelector;
        const selector =
          stateKey === 'disabled'
            ? `:${stateKey} ${nestedCssSelector || ''}`
            : `:${stateKey}:not(:disabled) ${nestedCssSelector || ''}`;
        const styles = getPresetStyles(presetState, options);
        if (stateKey === 'base') {
          if (nestedCssSelector) {
            acc[nestedCssSelector] = styles;
            return acc;
          }
          return {...acc, ...styles};
        }
        acc[selector] = styles;
      }

      return acc;
    },
    {} as CSSObject,
  );

export const getStylePresetFromTheme = <Props extends ThemeProp>(
  defaultToken?: MQ<string>,
  customProp?: Exclude<keyof Props, 'theme'>,
  options?: GetStylePresetFromThemeOptions,
) => (props: Props) => {
  const stylePreset = getResponsiveValueFromTheme('stylePresets')(
    defaultToken,
    customProp,
  )(props) as Partial<StylePreset> | Array<[string, StylePreset]>;

  if (Array.isArray(stylePreset)) {
    return stylePreset.reduce((acc, [mq, preset]) => {
      acc[mq] = getStylePresetValueFromTheme(preset, options);
      return acc;
    }, {} as CSSObject);
  }
  return stylePreset ? getStylePresetValueFromTheme(stylePreset, options) : '';
};

export const getStylePreset = getDefaultedValue(
  getStylePresetFromTheme,
  'stylePreset',
);

export const getSingleStylePreset = (
  {stylePresets}: Theme,
  state: StylePresetStates,
  cssProp: Exclude<StylePresetStyleKeys, 'borderRadius'>,
  defaultToken: string,
  customToken?: string,
) => {
  const preset =
    (customToken && stylePresets[customToken]) || stylePresets[defaultToken];
  return (preset[state] || {})[cssProp] || '';
};
