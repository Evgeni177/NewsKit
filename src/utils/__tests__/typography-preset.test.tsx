import {renderToFragmentWithTheme} from '../../test/test-utils';
import {getTypographyPresetFromTheme, styled, MQ} from '../style';

interface TestTextProp {
  typographyPreset: MQ<string>;
}

const TestText = styled.p<TestTextProp>`
  ${getTypographyPresetFromTheme('editorialParagraph030', 'typographyPreset')}
`;

describe('TypographyPreset', () => {
  test('renders default', () => {
    const fragment = renderToFragmentWithTheme(TestText);
    expect(fragment).toMatchSnapshot();
  });

  test('renders with override editorialParagraph010', () => {
    const fragment = renderToFragmentWithTheme(TestText, {
      typographyPreset: 'editorialParagraph010',
    });
    expect(fragment).toMatchSnapshot();
  });

  test('renders responsive override editorialParagraph010 for xs, editorialParagraph020 for sm, and editorialParagraph030 for md breakpoints', () => {
    const fragment = renderToFragmentWithTheme(TestText, {
      typographyPreset: {
        xs: 'editorialParagraph010',
        sm: 'editorialParagraph020',
        md: 'editorialParagraph030',
      },
    });
    expect(fragment).toMatchSnapshot();
  });

  test('renders responsive override editorialParagraph010 for xs, editorialParagraph020 for sm, and wrong for md breakpoints', () => {
    const fragment = renderToFragmentWithTheme(TestText, {
      typographyPreset: {
        xs: 'editorialParagraph010',
        sm: 'editorialParagraph020',
        md: 'wrong',
      } as any,
    });
    expect(fragment).toMatchSnapshot();
  });

  test('renders responsive override editorialParagraph010 for xs and editorialParagraph030 for md breakpoints', () => {
    const fragment = renderToFragmentWithTheme(TestText, {
      typographyPreset: {
        xs: 'editorialParagraph010',
        md: 'editorialParagraph030',
      },
    });
    expect(fragment).toMatchSnapshot();
  });

  test('renders responsive override editorialParagraph010 for xs, editorialParagraph020 for sm, editorialParagraph030 for md, editorialHeading010 for lg breakpoints and ignore wrong prop', () => {
    const fragment = renderToFragmentWithTheme(TestText, {
      typographyPreset: {
        xs: 'editorialParagraph010',
        sm: 'editorialParagraph020',
        md: 'editorialParagraph030',
        lg: 'editorialHeading010',
        wrong: 'editorialHeading020',
      } as any,
    });
    expect(fragment).toMatchSnapshot();
  });
});
