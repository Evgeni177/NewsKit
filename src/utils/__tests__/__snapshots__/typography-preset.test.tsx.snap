// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TypographyPreset renders default 1`] = `
<DocumentFragment>
  .emotion-0 {
  font-family: "Noto Sans",sans-serif;
  font-size: 18px;
  line-height: 1.5555555555555556;
  font-weight: 400;
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
}

<p
    class="emotion-0"
  />
</DocumentFragment>
`;

exports[`TypographyPreset renders responsive override editorialParagraph010 for xs and editorialParagraph030 for md breakpoints 1`] = `
<DocumentFragment>
  @media screen and (max-width:767px) {
  .emotion-0 {
    font-family: "Noto Sans",sans-serif;
    font-size: 14px;
    line-height: 1.4285714285714286;
    font-weight: 400;
    -webkit-letter-spacing: 0;
    -moz-letter-spacing: 0;
    -ms-letter-spacing: 0;
    letter-spacing: 0;
  }
}

@media screen and (min-width:768px) {
  .emotion-0 {
    font-family: "Noto Sans",sans-serif;
    font-size: 18px;
    line-height: 1.5555555555555556;
    font-weight: 400;
    -webkit-letter-spacing: 0;
    -moz-letter-spacing: 0;
    -ms-letter-spacing: 0;
    letter-spacing: 0;
  }
}

<p
    class="emotion-0"
  />
</DocumentFragment>
`;

exports[`TypographyPreset renders responsive override editorialParagraph010 for xs, editorialParagraph020 for sm, and editorialParagraph030 for md breakpoints 1`] = `
<DocumentFragment>
  @media screen and (max-width:479px) {
  .emotion-0 {
    font-family: "Noto Sans",sans-serif;
    font-size: 14px;
    line-height: 1.4285714285714286;
    font-weight: 400;
    -webkit-letter-spacing: 0;
    -moz-letter-spacing: 0;
    -ms-letter-spacing: 0;
    letter-spacing: 0;
  }
}

@media screen and (min-width:480px) and (max-width:767px) {
  .emotion-0 {
    font-family: "Noto Sans",sans-serif;
    font-size: 16px;
    line-height: 1.5;
    font-weight: 400;
    -webkit-letter-spacing: 0;
    -moz-letter-spacing: 0;
    -ms-letter-spacing: 0;
    letter-spacing: 0;
  }
}

@media screen and (min-width:768px) {
  .emotion-0 {
    font-family: "Noto Sans",sans-serif;
    font-size: 18px;
    line-height: 1.5555555555555556;
    font-weight: 400;
    -webkit-letter-spacing: 0;
    -moz-letter-spacing: 0;
    -ms-letter-spacing: 0;
    letter-spacing: 0;
  }
}

<p
    class="emotion-0"
  />
</DocumentFragment>
`;

exports[`TypographyPreset renders responsive override editorialParagraph010 for xs, editorialParagraph020 for sm, and wrong for md breakpoints 1`] = `
<DocumentFragment>
  @media screen and (max-width:479px) {
  .emotion-0 {
    font-family: "Noto Sans",sans-serif;
    font-size: 14px;
    line-height: 1.4285714285714286;
    font-weight: 400;
    -webkit-letter-spacing: 0;
    -moz-letter-spacing: 0;
    -ms-letter-spacing: 0;
    letter-spacing: 0;
  }
}

@media screen and (min-width:480px) {
  .emotion-0 {
    font-family: "Noto Sans",sans-serif;
    font-size: 16px;
    line-height: 1.5;
    font-weight: 400;
    -webkit-letter-spacing: 0;
    -moz-letter-spacing: 0;
    -ms-letter-spacing: 0;
    letter-spacing: 0;
  }
}

<p
    class="emotion-0"
  />
</DocumentFragment>
`;

exports[`TypographyPreset renders responsive override editorialParagraph010 for xs, editorialParagraph020 for sm, editorialParagraph030 for md, editorialHeading010 for lg breakpoints and ignore wrong prop 1`] = `
<DocumentFragment>
  @media screen and (max-width:479px) {
  .emotion-0 {
    font-family: "Noto Sans",sans-serif;
    font-size: 14px;
    line-height: 1.4285714285714286;
    font-weight: 400;
    -webkit-letter-spacing: 0;
    -moz-letter-spacing: 0;
    -ms-letter-spacing: 0;
    letter-spacing: 0;
  }
}

@media screen and (min-width:480px) and (max-width:767px) {
  .emotion-0 {
    font-family: "Noto Sans",sans-serif;
    font-size: 16px;
    line-height: 1.5;
    font-weight: 400;
    -webkit-letter-spacing: 0;
    -moz-letter-spacing: 0;
    -ms-letter-spacing: 0;
    letter-spacing: 0;
  }
}

@media screen and (min-width:768px) and (max-width:1023px) {
  .emotion-0 {
    font-family: "Noto Sans",sans-serif;
    font-size: 18px;
    line-height: 1.5555555555555556;
    font-weight: 400;
    -webkit-letter-spacing: 0;
    -moz-letter-spacing: 0;
    -ms-letter-spacing: 0;
    letter-spacing: 0;
  }
}

@media screen and (min-width:1024px) {
  .emotion-0 {
    font-family: "Source Serif Pro",serif;
    font-size: 18px;
    line-height: 1.1111111111111112;
    font-weight: 600;
    -webkit-letter-spacing: 0;
    -moz-letter-spacing: 0;
    -ms-letter-spacing: 0;
    letter-spacing: 0;
  }
}

<p
    class="emotion-0"
  />
</DocumentFragment>
`;

exports[`TypographyPreset renders with override editorialParagraph010 1`] = `
<DocumentFragment>
  .emotion-0 {
  font-family: "Noto Sans",sans-serif;
  font-size: 14px;
  line-height: 1.4285714285714286;
  font-weight: 400;
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
}

<p
    class="emotion-0"
  />
</DocumentFragment>
`;
